File: Problem1 Maximal Square DP.java
// TC : O(rows*cols)
// SC : O(row+1, cols+1) matrix

//IDEA : Add the minimum of up left and diagoanl left up valuse to curretn i + 1  in the DP array 
// Traverse the whole dp array and calculate the lenght of the square can be formend.
// return maxlength * maxlength

//LC Run : Success.


class Solution {
    public int maximalSquare(char[][] matrix) {
        int rows = matrix.length;
        if(rows == 0) return 0;
        int cols = matrix[0].length;
        
        int [][] dp = new int[rows+1][cols+1];
        int max =0;
        
        for(int i=1;i<=rows;i++){
            for(int j=1;j<=cols;j++){
                if(matrix[i-1][j-1] == '1')
                dp[i][j] =Math.min(Math.min(dp[i-1][j],dp[i][j-1]),dp[i-1][j-1]) +1;
                System.out.print(dp[i][j]+" ");
                max = Math.max(max,dp[i][j]);
            }
            System.out.print("\n ");
        }
        
        return max*max;
        
    }
}

File: Problem1_MaximalSquare_DP.py
'''
====== Submission Details =======
Student Name: Pavan Kumar K. N.
S30 SlackID : RN32MAY2021
=================================
'''

# 221. Maximal Square

# https://leetcode.com/problems/maximal-square/

#-----------------
# Time Complexity: 
#-----------------
# O( (M*N) ) - Where M is the number of rows and N is no. of columns in input

#------------------
# Space Complexity: 
#------------------
# O(N): Where N is the number of cols.

#-----------------------
# Leet Code Performance: 
#-----------------------
# Ran Successfully?: Yes


class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        m = len(matrix)
        if m > 0:
            n = len(matrix[0])
        else:
            n = 0
        
        max_val = 0
        
        dp = [0 for j in range(n+1)]
        
        #diagonal up-left element
        diag = 0

        for i in range(1, m+1):
            for j in range(1, n+1):
                if matrix[i-1][j-1] == "1":
                    temp = dp[j]
                    dp[j] = min(dp[j-1], dp[j], diag) + 1
                    max_val = max(max_val, dp[j])
                    diag = temp
                else:
                    dp[j] = 0
                
        
        return max_val**2
                        
                    
                        
                    